Source/Core/Core/IOS/USB/Emulated/Skylander.cpp

```diff
@@ -0,0 +1,717 @@
+{
+SkylanderPortal g_skyportal;
+
+SkylanderUSB::SkylanderUSB(Kernel& ios, const std::string& device_name) : m_ios(ios)
+{
+  m_vid = 0x1430;
+  m_pid = 0x150;
+  m_id = (static_cast<u64>(m_vid) << 32 | static_cast<u64>(m_pid) << 16 | static_cast<u64>(9) << 8 |
+          static_cast<u64>(1));
+  deviceDesc = DeviceDescriptor{18, 1, 512, 0, 0, 0, 64, 5168, 336, 256, 1, 2, 0, 1};
+  configDesc.emplace_back(ConfigDescriptor{9, 2, 41, 1, 1, 0, 128, 250});
+  interfaceDesc.emplace_back(InterfaceDescriptor{9, 4, 0, 0, 2, 3, 0, 0, 0});
+  endpointDesc.emplace_back(EndpointDescriptor{7, 5, 129, 3, 64, 1});
+  endpointDesc.emplace_back(EndpointDescriptor{7, 5, 2, 3, 64, 1});
+}
+
+SkylanderUSB::~SkylanderUSB()
+{
+}
+
+DeviceDescriptor SkylanderUSB::GetDeviceDescriptor() const
+{
+  return deviceDesc;
+}
+
+std::vector<ConfigDescriptor> SkylanderUSB::GetConfigurations() const
+{
+  return configDesc;
+}
+
+std::vector<InterfaceDescriptor> SkylanderUSB::GetInterfaces(u8 config) const
+{
+  return interfaceDesc;
+}
+
+std::vector<EndpointDescriptor> SkylanderUSB::GetEndpoints(u8 config, u8 interface, u8 alt) const
+{
+  return endpointDesc;
+}
+
+bool SkylanderUSB::Attach()
+{
+  if (m_device_attached)
+  {
+    return true;
+  }
+  DEBUG_LOG_FMT(IOS_USB, "[{:04x}:{:04x}] Opening device", m_vid, m_pid);
+  m_device_attached = true;
+  return true;
+}
+
+bool SkylanderUSB::AttachAndChangeInterface(const u8 interface)
+{
+  return true;
+}
+
+int SkylanderUSB::CancelTransfer(const u8 endpoint)
+{
+  INFO_LOG_FMT(IOS_USB, "[{:04x}:{:04x} {}] Cancelling transfers (endpoint {:#x})", m_vid, m_pid,
+               m_active_interface, endpoint);
+
+  return IPC_SUCCESS;
+}
+
+int SkylanderUSB::ChangeInterface(const u8 interface)
+{
+  DEBUG_LOG_FMT(IOS_USB, "[{:04x}:{:04x} {}] Changing interface to {}", m_vid, m_pid,
+                m_active_interface, interface);
+  m_active_interface = interface;
+  return 0;
+}
+
+int SkylanderUSB::GetNumberOfAltSettings(u8 interface)
+{
+  return 0;
+}
+
+int SkylanderUSB::SetAltSetting(u8 alt_setting)
+{
+  return 0;
+}
+
+// Skylander Portal control transfers are handled via this method, if a transfer requires a
+// response, then we request that data from the "Portal" and queue it to be responded to via the
+// Interrupt Response, references can be found via:
+// https://marijnkneppers.dev/posts
+// https://github.com/tresni/PoweredPortals/wiki/USB-Protocols
+// https://pastebin.com/EqtTRzeF
+int SkylanderUSB::SubmitTransfer(std::unique_ptr<CtrlMessage> cmd)
+{
+  DEBUG_LOG_FMT(IOS_USB,
+                "[{:04x}:{:04x} {}] Control: bRequestType={:02x} bRequest={} wValue={:04x}"
+                " wIndex={:04x} wLength={:04x}",
+                m_vid, m_pid, m_active_interface, cmd->request_type, cmd->request, cmd->value,
+                cmd->index, cmd->length);
+
+  auto& system = Core::System::GetInstance();
+  auto& memory = system.GetMemory();
+  u8* buf = memory.GetPointerForRange(cmd->data_address, cmd->length);
```

Comment:
You should check if you actually get a valid pointer here and return (with some error code) if you don't.

You should probably also error out if `cmd->length == 0`.